### PR TITLE
doc/rados: edit crush-map.rst (1 of x)

### PR DESCRIPTION
**Pull request by zdover23**
Edit doc/rados/operations/crush-map.rst.

https://tracker.ceph.com/issues/58485

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [x] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [x] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: doc/rados/operations/crush-map.rst [hunk 1]
```diff
@@ -3,103 +3,103 @@
...
+computes storage locations in order to determine how to store and retrieve
+data.  CRUSH allows Ceph clients to communicate with OSDs directly rather than
+through a centralized server or broker. By using an algorithmically-determined
 method of storing and retrieving data, Ceph avoids a single point of failure, a
 performance bottleneck, and a physical limit to its scalability.
 
-CRUSH uses a map of your cluster (the CRUSH map) to pseudo-randomly
-map data to OSDs, distributing it across the cluster according to configured
-replication policy and failure domain.  For a detailed discussion of CRUSH, see
+CRUSH uses a map of the cluster (the CRUSH map) to pseudorandomly map data to
```
Comment (anthonyeleven) [inline doc/rados/operations/crush-map.rst:12]: I'm not sure that `pseudorandomly` adds meaning here, I would consider dropping it.

File: doc/rados/operations/crush-map.rst [hunk 2]
```diff
@@ -3,103 +3,103 @@
...
+OSDs, distributing the data across the cluster with reference to configured
```
Comment (anthonyeleven) [inline doc/rados/operations/crush-map.rst:13]: s/with reference to/in accordance with/

File: doc/rados/operations/crush-map.rst [hunk 3]
```diff
@@ -3,103 +3,103 @@
...
+replication policy and failure domain. For a detailed discussion of CRUSH, see
```
Comment (anthonyeleven) [inline doc/rados/operations/crush-map.rst:14]: domains

File: doc/rados/operations/crush-map.rst [hunk 4]
```diff
@@ -3,103 +3,103 @@
...
-controllers, and/or physical locations.
-
-When you deploy OSDs they are automatically added to the CRUSH map under a
-``host`` bucket named for the node on which they run.  This,
-combined with the configured CRUSH failure domain, ensures that replicas or
-erasure code shards are distributed across hosts and that a single host or other
-failure will not affect availability.  For larger clusters, administrators must
-carefully consider their choice of failure domain.  Separating replicas across racks,
-for example, is typical for mid- to large-sized clusters.
+of 'buckets' for aggregating devices and S3 buckets, and rules that govern how
```
Comment (anthonyeleven) [inline doc/rados/operations/crush-map.rst:18]: Ceph is rife with overloaded nomenclature.

S3 has buckets that are more or less directories.  These are different.

The buckets here are CRUSH buckets -- OSDs, hosts, chassis, racks, rows, rooms, datacenters, etc.  Any type of node (there's another overload!) in the CRUSH hierarchy graph.

so I suggest

>list of OSDs and a hierarchy of "buckets" (``host``s, ``rack``s, etc.) and rules that govern how

File: doc/rados/operations/crush-map.rst [hunk 5]
```diff
@@ -3,103 +3,103 @@
...
+CRUSH replicates data within the cluster's pools. By reflecting the underlying
+physical organization of the installation, CRUSH can model (and thereby
+address) the potential for correlated device failures.  Some factors relevant
+to the CRUSH map include chassis, racks, physical proximity, a shared power
```
Comment (anthonyeleven) [inline doc/rados/operations/crush-map.rst:22]: s/map/hierarchy/ and failure domains/

File: doc/rados/operations/crush-map.rst [hunk 6]
```diff
@@ -3,103 +3,103 @@
...
+source, and shared networking. By encoding this information into the CRUSH map,
+CRUSH placement policies distribute object replicas across failure domains
+while maintaining the desired distribution. For example, to address the
+possibility of concurrent failures, it might be desirable to ensure that data
+replicas are on devices that use different shelves, racks, power supplies,
```
Comment (anthonyeleven) [inline doc/rados/operations/crush-map.rst:27]: s/use/reside in or rely upon/

File: doc/rados/operations/crush-map.rst [hunk 7]
```diff
@@ -3,103 +3,103 @@
...
+controllers, or physical locations.
+
+When OSDs are deployed, they are automatically added to the CRUSH map under a
+``host`` bucket that is named for the node on which the OSDs run. This
+behavior, combined with the configured CRUSH failure domain, ensures that
+replicas or erasure-code shards are distributed across hosts and that the
+failure of a single host or other kinds of failures will not affect
+availability. For larger clusters, administrators must carefully consider their
+choice of failure domain. For example, separating replicas across racks is
```
Comment (anthonyeleven) [inline doc/rados/operations/crush-map.rst:36]: s/separating/distributing/

File: doc/rados/operations/crush-map.rst [hunk 8]
```diff
@@ -3,103 +3,103 @@
...
 CRUSH Location
 ==============
 
-The location of an OSD within the CRUSH map's hierarchy is
-referred to as a ``CRUSH location``.  This location specifier takes the
-form of a list of key and value pairs.  For
-example, if an OSD is in a particular row, rack, chassis and host, and
-is part of the 'default' CRUSH root (which is the case for most
-clusters), its CRUSH location could be described as::
+The location of an OSD within the CRUSH map's hierarchy is referred to as a
```
Comment (anthonyeleven) [inline doc/rados/operations/crush-map.rst:43]: s/a/its/

File: doc/rados/operations/crush-map.rst [hunk 9]
```diff
@@ -108,34 +108,33 @@ based on a value in the file ``/etc/rack`` might be::
...
-Devices are identified by an ``id``
-(a non-negative integer) and a ``name``, normally ``osd.N`` where ``N`` is the device id.
+Devices are individual OSDs that store data (usually one device for each
+storage drive).  Devices are identified by an ``id`` (a non-negative integer)
+and a ``name`` (usually ``osd.N``, where ``N`` is the device's ``id``).
 
-Since the Luminous release, devices may also have a *device class* assigned (e.g.,
-``hdd`` or ``ssd`` or ``nvme``), allowing them to be conveniently targeted by
-CRUSH rules.  This is especially useful when mixing device types within hosts.
+In Luminous and later releases, devices can have a *device class* assigned (for
```
Comment (anthonyeleven) [inline doc/rados/operations/crush-map.rst:125]: s/devices/OSDs/

File: doc/rados/operations/crush-map.rst [hunk 10]
```diff
@@ -150,41 +149,42 @@ used to describe these nodes.  Default types include:
...
-weights are in units of terabytes (TB).
 
-You can get a simple view the of CRUSH hierarchy for your cluster,
-including weights, with:
+Each node (device or bucket) in the hierarchy has a *weight* that indicates the
+relative proportion of the total data that should be stored by that device or
+hierarchy subtree. Weights are set at the leaves, indicating the size of the
+device. These weights automatically sum in an 'up the tree' direction: that is,
+the weight of the ``root`` node will be the sum of the weights of all devices
+contained under it. Weights are typically measured in terabytes (TB).
```
Comment (anthonyeleven) [inline doc/rados/operations/crush-map.rst:184]: I think Tebibytes strictly, though the term sticks in my craw.

File: doc/rados/operations/crush-map.rst [hunk 11]
```diff
@@ -193,30 +193,28 @@ including weights, with:
...
-In rare cases rules must be written by hand by manually editing the
-CRUSH map.
-
-You can see what rules are defined for your cluster with:
+CRUSH rules define policy governing how data is distributed across the devices
+in the hierarchy. The rules define placement as well as replication strategies
+or distribution policies that allow you to specify exactly how CRUSH places
+data replicas. For example, you might create one rule selecting a pair of
+targets for two-way mirroring, another rule for selecting three targets in two
+different data centers for three-way mirroring, and yet another rule for
```
Comment (anthonyeleven) [inline doc/rados/operations/crush-map.rst:201]: 2x s/mirroring/replication/

File: doc/rados/operations/crush-map.rst [hunk 12]
```diff
@@ -193,30 +193,28 @@ including weights, with:
...
+erasure coding across six storage devices. For a detailed discussion of CRUSH
+rules, see **Section 3.2** of `CRUSH - Controlled, Scalable, Decentralized
+Placement of Replicated Data`_.
+
+CRUSH rules can be created via the command-line interface by specifying the
```
Comment (anthonyeleven) [inline doc/rados/operations/crush-map.rst:206]: s/interface//

File: doc/rados/operations/crush-map.rst [hunk 13]
```diff
@@ -225,106 +223,106 @@ You can view the contents of the rules with:
 Device classes
 --------------
 
-Each device can optionally have a *class* assigned.  By
-default, OSDs automatically set their class at startup to
-`hdd`, `ssd`, or `nvme` based on the type of device they are backed
-by.
+Each device can optionally have a *class* assigned. By default, OSDs
+automatically set their class at startup to `hdd`, `ssd`, or `nvme` in
+accordance with the type of device they are backed by.
```
Comment (anthonyeleven) [inline doc/rados/operations/crush-map.rst:228]: If `nvme` is automatically set, that's new -- NVMe devices in my experience show up as `ssd`.  But let's leave this as-is for now.

**Review by anthonyeleven — APPROVED** [review]
Back to you

File: doc/rados/operations/crush-map.rst (hunk 1 above)
Comment (zdover23) [reply doc/rados/operations/crush-map.rst:12]: Accepted.

File: doc/rados/operations/crush-map.rst (hunk 2 above)
Comment (zdover23) [reply doc/rados/operations/crush-map.rst:13]: Accepted.

File: doc/rados/operations/crush-map.rst (hunk 4 above)
Comment (zdover23) [reply doc/rados/operations/crush-map.rst:18]: Accepted.

File: doc/rados/operations/crush-map.rst (hunk 5 above)
Comment (zdover23) [reply doc/rados/operations/crush-map.rst:22]: Accepted.

File: doc/rados/operations/crush-map.rst (hunk 6 above)
Comment (zdover23) [reply doc/rados/operations/crush-map.rst:27]: Accepted.

File: doc/rados/operations/crush-map.rst (hunk 7 above)
Comment (zdover23) [reply doc/rados/operations/crush-map.rst:36]: Accepted.

File: doc/rados/operations/crush-map.rst (hunk 8 above)
Comment (zdover23) [reply doc/rados/operations/crush-map.rst:43]: Accepted.

File: doc/rados/operations/crush-map.rst (hunk 9 above)
Comment (zdover23) [reply doc/rados/operations/crush-map.rst:125]: Accepted.

File: doc/rados/operations/crush-map.rst (hunk 10 above)
Comment (zdover23) [reply doc/rados/operations/crush-map.rst:184]: Accepted.

File: doc/rados/operations/crush-map.rst (hunk 12 above)
Comment (zdover23) [reply doc/rados/operations/crush-map.rst:206]: Accepted.

**Comment by zdover23** [comment]
https://github.com/ceph/ceph/pull/52030 - Reef backport
https://github.com/ceph/ceph/pull/52031 - Quincy backport

**Pull request merged**